### PR TITLE
provide a way to plug-in an external authorizer

### PR DESCRIPTION
**Pull request by pureneelesh**
**What this PR does / why we need it**:

Added framework for plugging in an external authorizer such as an OPA.

Signed-off-by: Neelesh Thakur <neelesh.thakur@purestorage.com>

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

File: server/grpc_framework_server.go [hunk 1]
```diff
@@ -73,8 +73,13 @@ func NewGrpcFrameworkServer(config *ServerConfig) (*GrpcFrameworkServer, error)
...
-		// Check the necessary security config options are set
-		if config.Security.Role == nil && (config.AuthZUnaryInterceptor == nil || config.AuthZStreamInterceptor == nil) {
+		// Check the necessary security config options are set. Authentication is enabled. Therefore,
+		// either RoleManager must be provided (this implies use of the default authZ)
+		// or explicit authZ interceptors must be provided
+		// or external authZ checker must be proviced (this implies use of external_authorizer.go)
+		if config.Security.Role == nil &&
+			(config.AuthZUnaryInterceptor == nil || config.AuthZStreamInterceptor == nil) &&
+			config.ExternalAuthZChecker == nil {
 			return nil, fmt.Errorf("must supply role manager when authentication is enabled and default authZ is used")
```
Comment (adityadani) [inline server/grpc_framework_server.go:83]: change the error message?

File: server/external_authorizer.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		}
+		authZReq, handlerData, err := authZReqGetter.GetAuthZRequest(ctx, info.FullMethod, apiRequest)
+		if err != nil {
+			return nil, auditLogErrorf(codes.Internal, "Failed to get authZ request: %v", err)
+		}
+		switch authZReq {
+		case InsecureNoAuthNAuthZ:
+			// skip authZ check as long as the request type exists in our list
+			if !s.listContainsReqType(insecureNoAuthNAuthZReqs, apiRequest) {
+				return nil, auditLogErrorf(codes.Internal, "req %T absent in skip authZ list", apiRequest)
```
Comment (adityadani) [inline server/external_authorizer.go:111]: nit: absent in skip authN and authZ list

File: server/external_authorizer.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			}
+		case InsecureNoAuthZ:
+			if !s.listContainsReqType(insecureNoAuthZReqs, apiRequest) {
+				return nil, auditLogErrorf(codes.Internal, "req %T absent in allow authenticated list", apiRequest)
```
Comment (adityadani) [inline server/external_authorizer.go:115]: nit: absent in skip authZ list

File: server/external_authorizer.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+			}
+			// any authenticated user is allowed
+			if !userAuthenticated {
+				return nil, auditLogErrorf(codes.Unauthenticated, "Authentication creds not found")
```
Comment (adityadani) [inline server/external_authorizer.go:119]: nit: should all these errors start with lower case?

**Review by adityadani — APPROVED** [review]
lgtm

File: server/external_authorizer.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+type insecureNoAuthNAuthZ struct{}
+
+// InsecureNoAuthNAuthZ is returned by the API handlers that wish to skip authN and authZ checks.
+var InsecureNoAuthNAuthZ insecureNoAuthNAuthZ
+
+type insecureNoAuthZ struct{}
+
+// InsecureNoAuthZ is returned by the API handlers that want to skip authZ but require that
+// the user be an authenticated user. No guests.
+var InsecureNoAuthZ insecureNoAuthZ
```
Comment (piyush-nimbalkar) [inline server/external_authorizer.go:34]: Any reason why these can't be constants of a single type?

File: server/grpc_framework_server.go (hunk 1 above)
Comment (pureneelesh) [reply server/grpc_framework_server.go:83]: @adityadani : Seems like the error message is still correct. What would you like to be changed?

**Review by piyush-nimbalkar — APPROVED** [review]

**Pull request merged**